src/syscall/libc_wasip2.go

```diff
@@ -0,0 +1,1294 @@
+	// with a '/'. This is needed to ensure that programs have a mechanism
+	// to represent dereferencing symbolic links pointing to directories.
+	if buf[len(buf)-1] != '/' && isDir(file) {
+		buf = append(buf, '/')
+	}
+	return unsafe.String(&buf[0], len(buf))
+}
+
+func isAbs(path string) bool {
+	return hasPrefix(path, "/")
+}
+
+func isDir(path string) bool {
+	return hasSuffix(path, "/")
+}
+
+func hasPrefix(s, p string) bool {
+	return len(s) >= len(p) && s[:len(p)] == p
+}
+
+func hasSuffix(s, x string) bool {
+	return len(s) >= len(x) && s[len(s)-len(x):] == x
+}
+
+// findPreopenForPath finds which preopen it relates to and return that descriptor/root and the path relative to that directory descriptor/root
+func findPreopenForPath(path string) (wasiDir, string) {
+	dir := "/"
+	var wasidir wasiDir
+
+	if !isAbs(path) {
+		dir = libcCWD.root
+		wasidir = libcCWD
+		if libcCWD.rel != "" && libcCWD.rel != "." && libcCWD.rel != "./" {
+			path = libcCWD.rel + "/" + path
+		}
+	}
+	path = joinPath(dir, path)
+
+	var best string
+	for k, v := range wasiPreopens {
+		if len(k) > len(best) && hasPrefix(path, k) {
+			wasidir = wasiDir{d: v, root: k}
+			best = wasidir.root
+		}
+	}
+
+	if hasPrefix(path, wasidir.root) {
+		path = path[len(wasidir.root):]
+	}
+	for isAbs(path) {
+		path = path[1:]
+	}
+	if len(path) == 0 {
+		path = "."
+	}
+
+	return wasidir, path
+}
+
+// int open(const char *pathname, int flags, mode_t mode);
+//
+//go:export open
+func open(pathname *byte, flags int32, mode uint32) int32 {
+	path := goString(pathname)
+	dir, relPath := findPreopenForPath(path)
+	if dir.d == cm.ResourceNone {
+		libcErrno = EACCES
+		return -1
+	}
+
+	var dflags types.DescriptorFlags
+	if (flags & O_RDONLY) == O_RDONLY {
+		dflags |= types.DescriptorFlagsRead
+	}
+	if (flags & O_WRONLY) == O_WRONLY {
+		dflags |= types.DescriptorFlagsWrite
+	}
+
+	var oflags types.OpenFlags
+	if flags&O_CREAT == O_CREAT {
+		oflags |= types.OpenFlagsCreate
+	}
+	if flags&O_DIRECTORY == O_DIRECTORY {
+		oflags |= types.OpenFlagsDirectory
+	}
+	if flags&O_EXCL == O_EXCL {
+		oflags |= types.OpenFlagsExclusive
+	}
+	if flags&O_TRUNC == O_TRUNC {
+		oflags |= types.OpenFlagsTruncate
+	}
+
+	// By default, follow symlinks for open() unless O_NOFOLLOW was passed
+	var pflags types.PathFlags = types.PathFlagsSymlinkFollow
+	if flags&O_NOFOLLOW == O_NOFOLLOW {
+		// O_NOFOLLOW was passed, so turn off SymlinkFollow
+		pflags &^= types.PathFlagsSymlinkFollow
+	}
+
+	result := dir.d.OpenAt(pflags, relPath, oflags, dflags)
+	if err := result.Err(); err != nil {
+		libcErrno = errorCodeToErrno(*err)
+		return -1
+	}
+
+	stream := wasiFile{
+		d:     *result.OK(),
+		oflag: flags,
+	}
+
+	if flags&(O_WRONLY|O_APPEND) == (O_WRONLY | O_APPEND) {
+		result := stream.d.Stat()
+		if err := result.Err(); err != nil {
+			libcErrno = errorCodeToErrno(*err)
+			return -1
+		}
+		stream.offset = int64(result.OK().Size)
+	}
+
+	libcfd := nextLibcFd
+	nextLibcFd++
```

Comment:
Shouldn't this use the `findFreeFd()` function? Or am I missing something?

Reply:
Refcounting added and tested.